RFC: dedicated storage engine for Raft

irfansharif opened the pull request:
Each `Replica` is backed by a single instance of RocksDB which is used to
store all modifications to the underlying state machine _in addition_ to
storing all consensus state.  This RFC proposes the separation of the
two, outlines the motivations for doing so and alternatives considered.

irfansharif committed:
RFC: dedicated storage engine for Raft

Each Replica is backed by a single instance of RocksDB which is used to
store all modifications to the underlying state machine in addition to
storing all consensus state. This RFC proposes the separation of the
two, outlines the motivations for doing so and alternatives considered.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16361)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice RFC!

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-KlyeVG2P82S8WB3WuSZ:by9nc26) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
> ```Markdown
> 
> ```go
> func BenchmarkBatchCommitSequential(b *testing.B) {
> ```

I can see the value in this benchmark, but I'd be happier with an alternating benchmark that used 2 separate RocksDB instances. Hopefully this would get similar numbers to this benchmark.

---

*[docs/RFCS/dedicated_raft_storage.md, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyg8cuXzGTcLhVpF5y:-Klyg8cuXzGTcLhVpF5z:bk1cilv) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L241)):*
> ```Markdown
> done for all reads, we will stop writing out raft entries to the original
> RocksDB instance.  Raft specific reads, writes and log truncations will now be
> serviced by the new instance. </br>
> ```

Since we're punting on the migration story for now, it seems worthwhile to structure this so that both a shared RocksDB instance and a separate RocksDB instance can be used, controlled by an env var. 

Also, given that the goal in this RFC is a performance improvement, it would be good to structure work so that we can get a full-system sanity check of the performance soon, even if this means doing something slightly hacky.

---

*[docs/RFCS/dedicated_raft_storage.md, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyf7PWNidttYPsm2x-:-Klyf7PWNidttYPsm2x0:b-kbnc1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L259)):*
> ```Markdown
> 
> TBD. The Raft log is tightly coupled to our consistency checker, this will be
> something to be wary of in order to transition safely.
> ```

Also TBD is how to actually start using this RocksDB instance. One thought is that only new nodes would use a separate RocksDB instance for the Raft log. Another is to add a store-level migration. Let's not explore these too deeply yet, just mention them briefly.

---

*[docs/RFCS/dedicated_raft_storage.md, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyfVut6CdkTkP1ECAr:-KlyfVut6CdkTkP1ECAs:b-3coybj) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L292)):*
> ```Markdown
> for the Raft log usage patterns. Possible reasons for doing so:
> - A native implementation in Go would avoid the CGo overhead we incur crossing
>   the G/C++ boundary
> ```

`s/G/Go/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-:-Klyn3X0gTQiPQU6if5E:b-46zc62)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great writeup, @irfansharif

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-Klz-ay3PYoezEaf-Bf-:b-a7rwsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I can see the value in this benchmark, but I'd be happier with an alternating benchmark that used 2 separate RocksDB instances. Hopefully this would get similar numbers to this benchmark.
</blockquote></details>

Yeah, I was going to say the same thing. There might be some interference between the two that would be very relevant to the accuracy of the benchmark.

---

*[docs/RFCS/dedicated_raft_storage.md, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz-HUd27QOU-U0bFhA:-Klz-HUd27QOU-U0bFhB:bolaxgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L81)):*
> ```Markdown
>     for _, valueSize := range []int{1 << 10, 1 << 12, ..., 1 << 20} {
> 		b.Run(fmt.Sprintf("vs=%d", valueSize), func(b *testing.B) {
>             // ...
> ```

You may want to clean up the indentation in here to make it render properly

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz0qzvxmntIPpPg0uD:-Klz0qzvxmntIPpPg0uE:brpswp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L247)):*
> RocksDB by default uses faster `fdatasync()` to sync files. We'll need to
> use `fsync()` instead in filesystems like ext3 where you can lose files after a
> reboot, this can be done so by setting `Options::use_fsync` to `true` (for the
> `WriteBatch`).

This comment is kinda coming out of the blue. Feel free to punt these questions out into a separate issue, but does this problem affect us today? Will we always enable `use_fsync`? If not, how will we tell when we need to?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-:-Klz2RLypRvWEOAaSSqs:b-uujlt3)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyg8cuXzGTcLhVpF5y:-Klz8dJt88jTc8fzWdlh:b-m1oz4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L241)):*
>  it would be good to structure work so that we can get a full-system sanity check of the performance soon, even if this means doing something slightly hacky.

I'm not sure I follow, do you mind clarifying?

---

*[docs/RFCS/dedicated_raft_storage.md, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz-HUd27QOU-U0bFhA:-Klz3zwKP1ZFYRzPTe-u:bk0f2o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L81)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You may want to clean up the indentation in here to make it render properly
</blockquote></details>

hmm, strange. works for me™:
![✂-1](https://files.reviewableusercontent.io/images/reviewable/10536690/DQQ-04vexPTRnJExAMH-RPHR.png)

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz0qzvxmntIPpPg0uD:-Klz4KniGU2mKSsGh0kb:bivbrf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L247)):*
Source: [rocksdb/wiki](https://github.com/facebook/rocksdb/wiki/basic-operations#asynchronous-writes). 

> does this problem affect us today? 

possibly, yes.

> Will we always enable `use_fsync`? If not, how will we tell when we need to?

like mentioned above, only for `ext3` filesystems. This is a flag set on `rocksdb::Options`, it _should_ happen [here](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/storage/engine/db.cc#L1490) but doesn't seem so. As for how, I'm sure there's a way to detect if the instance is configured to run on an `ext3` filesystem.

This jumped out to me as well, I couldn't find instances of us accounting for this fact but I'll file + address separately once I confirm. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/dedicated_raft_storage.md, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz-HUd27QOU-U0bFhA:-KlzNzoCb911dpkLMCr5:bj55gkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L81)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

hmm, strange. works for me™:
![✂-1](https://files.reviewableusercontent.io/images/reviewable/10536690/DQQ-04vexPTRnJExAMH-RPHR.png)

</blockquote></details>

Not on github in chrome:

![Screen Shot 2017-06-06 at 5.51.06 PM.png](https://files.reviewableusercontent.io/images/reviewable/7085343/FusP-I3QP3tCO4I7yCNzTY7T/Screen%20Shot%202017-06-06%20at%205.51.06%20PM.png)

Reviewable is even indicating that there's some whitespace weirdness in that some of the lines have the red >> arrows and others don't.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyg8cuXzGTcLhVpF5y:-Klze2UEL-JdOIZeZnMu:b-vabd72) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L241)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

>  it would be good to structure work so that we can get a full-system sanity check of the performance soon, even if this means doing something slightly hacky.

I'm not sure I follow, do you mind clarifying?
</blockquote></details>

Added these notes to the RFC.

---

*[docs/RFCS/dedicated_raft_storage.md, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyf7PWNidttYPsm2x-:-Klze60f7ERLeAbtvEli:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L259)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Also TBD is how to actually start using this RocksDB instance. One thought is that only new nodes would use a separate RocksDB instance for the Raft log. Another is to add a store-level migration. Let's not explore these too deeply yet, just mention them briefly.
</blockquote></details>

Done.

---

*[docs/RFCS/dedicated_raft_storage.md, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klz-HUd27QOU-U0bFhA:-KlzROTXG4xkIFTsHNNW:b4e0qox) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7cb3c5d3cd45189dd8ec97625bd9e72e4e91903/docs/RFCS/dedicated_raft_storage.md#L81)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Not on github in chrome:

![Screen Shot 2017-06-06 at 5.51.06 PM.png](https://files.reviewableusercontent.io/images/reviewable/7085343/FusP-I3QP3tCO4I7yCNzTY7T/Screen%20Shot%202017-06-06%20at%205.51.06%20PM.png)

Reviewable is even indicating that there's some whitespace weirdness in that some of the lines have the red >> arrows and others don't.
</blockquote></details>

strange, Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 272 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XzL1vuFDy5cLOso8:-Km-XzL1vuFDy5cLOso9:b-g9et2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L272)):*
> ```Markdown
> TBD. Some things to take note of:
> - The Raft log is tightly coupled to our consistency checker, this will be
>   something to be wary of in order to transition safely.
> ```

I'm not seeing the issue here. The consistency checker doesn't check the raft log itself. All we require is that the consistency checker gets the right snapshot of the main database, which is controlled by the way we apply entries and track the AppliedIndex in the main database (not the raft one)

---

*[docs/RFCS/dedicated_raft_storage.md, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-Xk60XYD3fK3NriGF:-Km-Xk60XYD3fK3NriGG:brbj0uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L274)):*
> ```Markdown
>   something to be wary of in order to transition safely.
> - How do we actually start using this RocksDB instance? One thought is that only
>   new nodes would use a separate RocksDB instance for the Raft log. Another
> ```

I'd prefer a general migration story instead of a one-time "new nodes use the new format" migration. In addition to solving the problem for existing clusters, a general migration process could be used to move the raft rocksdb instance from one location to another (such as to the non-volatile memory discussed below)

---

*[docs/RFCS/dedicated_raft_storage.md, line 284 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XOVoei9jnvpnWpgE:-Km-XOVoei9jnvpnWpgF:b-md5pfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L284)):*
> ```Markdown
> 
> It is not immediately obvious how much disk space should be allocated for the
> Raft specific RocksDB engine. This is not something we had to concern ourselves
> ```

If we weren't concerned about it before, why should we be concerned about it now? It doesn't seem like anything has changed. RocksDB's management of the disk space doesn't change just because there are two instances. (The raft logs compete for space with the regular data just as much today). 

---

*[docs/RFCS/dedicated_raft_storage.md, line 378 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XA6xEm35-9oFOrxn:-Km-XA6xEm35-9oFOrxo:b-rj0zbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L378)):*
> ```Markdown
> in non-volatile memory instead of disk (for etcd/raft)<sup>[6]</sup>.
> Given we're proposing a separate storage engine for the Raft log, in the
> presence of more suitable hardware medium it should be easy enough to configure
> ```

Even without more specialized hardware it might be desirable to configure the raft rocksdb and regular rocksdb to use different disks. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-:-Km-YO8dmOQyoWRDBWjR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Klyg8cuXzGTcLhVpF5y:-Km1SJRnSfivTBgCVM4A:bcb7omr) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L241)):*
> I'm not sure I follow, do you mind clarifying?

Build it so that we can benchmark it ASAP, even if that benchmarked version has blemishes or unaddressed problems that are not expected to influence benchmarks. Just to avoid finding out at the end that the numbers are not as favorable after having put in a lot of work.

---

*[docs/RFCS/dedicated_raft_storage.md, line 57 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km1UA2Yxd_TjPxKgj_3:-Km1UA2Yxd_TjPxKgj_4:bkpy6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L57)):*
> ```Markdown
> alternative below).
> 
> By having a dedicated storage engine for Raft's persistent state we can address
> ```

Can you briefly list which keys actually go in the new engine? Off the top of my head I'm thinking `HardState` and log keys, but would be good to have a list here.

---

*[docs/RFCS/dedicated_raft_storage.md, line 256 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km1SeIA5igPQNw48Sch:-Km1SeIBz7qIqkc3Lavk:bxbnwfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L256)):*
> ```Markdown
> **NB**: There's a subtle edge case to be wary of with respect to raft log
> truncations, before truncating the raft log we need to ensure that the
> application of the truncated entries have actually been persisted.</br>
> ```

You're worried that you would purge the log entries, but then not commit the updated `TruncatedState`, right? This command is going to be interesting anyways because I *think* it wants to write to both engines. Think this is worth expanding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-:-Km1VETT2GqSVW7uZCtv:bnfp4nl)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-Km8lKhoy8Dji4Ce0RT9:b3hu2ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Yeah, I was going to say the same thing. There might be some interference between the two that would be very relevant to the accuracy of the benchmark.
</blockquote></details>

hmm, I was looking at this today and tried out the alternating benchmark idea using 2 separate RocksDB instances, it's _mostly_ what I expect (similar to results posted here) save for value sizes of 512 KiB (65536 bytes).

```
~ benchstat perf-old.txt perf-new.txt
name                      old time/op    new time/op    delta
BatchCommit/vs=1024-4       74.0µs ± 3%    71.5µs ± 2%   -3.39%  (p=0.000 n=17+17)
BatchCommit/vs=4096-4        120µs ± 9%     109µs ±16%   -9.18%  (p=0.000 n=20+20)
BatchCommit/vs=16384-4       324µs ±11%     213µs ± 2%  -34.24%  (p=0.000 n=20+18)
BatchCommit/vs=65536-4      1.13ms ±33%    1.13ms ±10%     ~     (p=0.214 n=20+19)
BatchCommit/vs=262144-4     3.63ms ± 4%    2.92ms ± 4%  -19.64%  (p=0.000 n=20+20)
BatchCommit/vs=1048576-4    10.4ms ± 5%     8.1ms ± 5%  -21.97%  (p=0.000 n=16+20)

name                      old speed      new speed      delta
BatchCommit/vs=1024-4     13.8MB/s ± 3%  14.3MB/s ± 2%   +3.50%  (p=0.000 n=17+17)
BatchCommit/vs=4096-4     34.1MB/s ± 8%  37.6MB/s ±14%  +10.31%  (p=0.000 n=20+20)
BatchCommit/vs=16384-4    50.6MB/s ±10%  76.9MB/s ± 2%  +51.81%  (p=0.000 n=20+18)
BatchCommit/vs=65536-4    58.5MB/s ±26%  57.9MB/s ±10%     ~     (p=0.214 n=20+19)
BatchCommit/vs=262144-4   72.2MB/s ± 4%  89.8MB/s ± 4%  +24.42%  (p=0.000 n=20+20)
BatchCommit/vs=1048576-4   101MB/s ± 5%   130MB/s ± 5%  +28.17%  (p=0.000 n=16+20)

```

The actual code is [here](https://gist.github.com/irfansharif/3d710beea17655c41f95741d51367cf0). Any ideas for what's so special about 512 KiB?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-Km9Ew9CFT2cC8Lg2bLf:b-tm4gos) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

hmm, I was looking at this today and tried out the alternating benchmark idea using 2 separate RocksDB instances, it's _mostly_ what I expect (similar to results posted here) save for value sizes of 512 KiB (65536 bytes).

```
~ benchstat perf-old.txt perf-new.txt
name                      old time/op    new time/op    delta
BatchCommit/vs=1024-4       74.0µs ± 3%    71.5µs ± 2%   -3.39%  (p=0.000 n=17+17)
BatchCommit/vs=4096-4        120µs ± 9%     109µs ±16%   -9.18%  (p=0.000 n=20+20)
BatchCommit/vs=16384-4       324µs ±11%     213µs ± 2%  -34.24%  (p=0.000 n=20+18)
BatchCommit/vs=65536-4      1.13ms ±33%    1.13ms ±10%     ~     (p=0.214 n=20+19)
BatchCommit/vs=262144-4     3.63ms ± 4%    2.92ms ± 4%  -19.64%  (p=0.000 n=20+20)
BatchCommit/vs=1048576-4    10.4ms ± 5%     8.1ms ± 5%  -21.97%  (p=0.000 n=16+20)

name                      old speed      new speed      delta
BatchCommit/vs=1024-4     13.8MB/s ± 3%  14.3MB/s ± 2%   +3.50%  (p=0.000 n=17+17)
BatchCommit/vs=4096-4     34.1MB/s ± 8%  37.6MB/s ±14%  +10.31%  (p=0.000 n=20+20)
BatchCommit/vs=16384-4    50.6MB/s ±10%  76.9MB/s ± 2%  +51.81%  (p=0.000 n=20+18)
BatchCommit/vs=65536-4    58.5MB/s ±26%  57.9MB/s ±10%     ~     (p=0.214 n=20+19)
BatchCommit/vs=262144-4   72.2MB/s ± 4%  89.8MB/s ± 4%  +24.42%  (p=0.000 n=20+20)
BatchCommit/vs=1048576-4   101MB/s ± 5%   130MB/s ± 5%  +28.17%  (p=0.000 n=16+20)

```

The actual code is [here](https://gist.github.com/irfansharif/3d710beea17655c41f95741d51367cf0). Any ideas for what's so special about 512 KiB?
</blockquote></details>

The benchmark looks good. The `vs=65536` runs have high variance. Were you doing anything else on the machine while running the benchmarks? Even browsing the web can foul things up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 57 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km1UA2Yxd_TjPxKgj_3:-KmCfJHPgZCjue6ObqXd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L57)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can you briefly list which keys actually go in the new engine? Off the top of my head I'm thinking `HardState` and log keys, but would be good to have a list here.
</blockquote></details>

Done.

---

*[docs/RFCS/dedicated_raft_storage.md, line 256 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km1SeIA5igPQNw48Sch:-KmCfGmfelexlA7xKPS5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L256)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You're worried that you would purge the log entries, but then not commit the updated `TruncatedState`, right? This command is going to be interesting anyways because I *think* it wants to write to both engines. Think this is worth expanding.
</blockquote></details>

Done.

---

*[docs/RFCS/dedicated_raft_storage.md, line 272 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XzL1vuFDy5cLOso8:-KmCjtiIDB_LqxuANDbC:b-nog9vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L272)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm not seeing the issue here. The consistency checker doesn't check the raft log itself. All we require is that the consistency checker gets the right snapshot of the main database, which is controlled by the way we apply entries and track the AppliedIndex in the main database (not the raft one)
</blockquote></details>

ah, I was conflating `TruncationState` being coupled to `ReplicaState` (thereby not allowing for truncations to happen independently across replicas). Removed.

---

*[docs/RFCS/dedicated_raft_storage.md, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-Xk60XYD3fK3NriGF:-KmDSqAEChr7QIRcqE9u:bx6w0rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L274)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd prefer a general migration story instead of a one-time "new nodes use the new format" migration. In addition to solving the problem for existing clusters, a general migration process could be used to move the raft rocksdb instance from one location to another (such as to the non-volatile memory discussed below)
</blockquote></details>

I've added some thoughts I have on this, PTAL as I've found little precedence for this in the past.

---

*[docs/RFCS/dedicated_raft_storage.md, line 284 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XOVoei9jnvpnWpgE:-KmDB4RNMtqmxw3geHja:b-1akpun) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L284)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If we weren't concerned about it before, why should we be concerned about it now? It doesn't seem like anything has changed. RocksDB's management of the disk space doesn't change just because there are two instances. (The raft logs compete for space with the regular data just as much today). 
</blockquote></details>

I was under the impression that each RocksDB instance was bounded by the `maxSize` it was initialized with (`storage/engine/rocksdb.go: NewRocksDB`). Assuming each instance has to be initialized with it's own `maxSize`, knowing ahead of time what ratio of total available disk space gets assigned to each instance is not obvious as yet. The point about investigating multiple embedded RocksDB databases was to this end, i.e. if somehow both instances can be configured to compete for the same total space as before.

---

*[docs/RFCS/dedicated_raft_storage.md, line 378 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XA6xEm35-9oFOrxn:-KmDDOyIIh6ng9gUdf4C:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L378)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Even without more specialized hardware it might be desirable to configure the raft rocksdb and regular rocksdb to use different disks. 
</blockquote></details>

Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-KmDUGGaxL7I14GfXeY6:b-svma60) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

nope, and I've gotten similar results across multiple runs of the same experiment. Poking around this I found something else that was curious.
[Here](https://gist.github.com/irfansharif/70d138e638600693935c924be5df8178) are two versions of a benchmark, only difference being that in one I have a single loop body and write out `b.N` synced & unsynced writes (interleaved) vs. two loop bodies with `b.N` synced & unsynced writes each.
As before they're directed at separately running instances, here are the perf differences (especially stark for 64 KiB workloads):
```
~ benchstat perf-interleaved.txt perf-sequential.txt
name                      old time/op    new time/op     delta
BatchCommit/vs=1024-4       70.1µs ± 2%     68.6µs ± 5%   -2.15%  (p=0.021 n=8+10)
BatchCommit/vs=4096-4        102µs ± 1%       97µs ± 7%   -4.10%  (p=0.013 n=9+10)
BatchCommit/vs=16384-4       207µs ± 5%      188µs ± 4%   -9.46%  (p=0.000 n=9+9)
BatchCommit/vs=65536-4      1.07ms ±12%     0.62ms ± 9%  -41.90%  (p=0.000 n=8+9)
BatchCommit/vs=262144-4     2.90ms ± 8%     2.70ms ± 4%   -6.68%  (p=0.000 n=9+10)
BatchCommit/vs=1048576-4    8.06ms ± 9%     7.90ms ± 5%     ~     (p=0.631 n=10+10)

name                      old speed      new speed       delta
BatchCommit/vs=1024-4     14.6MB/s ± 2%   14.9MB/s ± 4%   +2.22%  (p=0.021 n=8+10)
BatchCommit/vs=4096-4     40.3MB/s ± 1%   42.1MB/s ± 7%   +4.37%  (p=0.013 n=9+10)
BatchCommit/vs=16384-4    78.6MB/s ± 5%   87.4MB/s ± 3%  +11.09%  (p=0.000 n=10+9)
BatchCommit/vs=65536-4    61.6MB/s ±13%  105.5MB/s ± 8%  +71.32%  (p=0.000 n=8+9)
BatchCommit/vs=262144-4   90.6MB/s ± 7%   97.1MB/s ± 4%   +7.13%  (p=0.000 n=9+10)
BatchCommit/vs=1048576-4   130MB/s ± 8%    133MB/s ± 5%     ~     (p=0.631 n=10+10)

```

Only reasonable conclusion I can draw from this is that the separately running instances
are not as isolated as expected. Given it's running on a single disk machine, possible something
to do with disk and/or OS buffers? I can update the RFC with these results (i.e. skewed 64 KiB 
workloads) and make a note to investigate further down the line.
I also tried initializing each instance with it's separately sized cache and upper disk usage limit
(changed `setupMVCCRocksDB` to this end to not use defaults) but to no effect.

</blockquote></details>

Given my recent work on https://github.com/cockroachdb/cockroach/issues/14108, we'll definitely have to pay attention to how the separate RocksDB instances interact. For example, I seem to recall that we can configure them to use a shared background thread pool. Would that be better, or should they have separate thread pools for background compactions? 

---

*[docs/RFCS/dedicated_raft_storage.md, line 284 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XOVoei9jnvpnWpgE:-KmDWNoFMiG2GU0H3gq9:bldtvq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L284)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I was under the impression that each RocksDB instance was bounded by the `maxSize` it was initialized with (`storage/engine/rocksdb.go: NewRocksDB`). Assuming each instance has to be initialized with it's own `maxSize`, knowing ahead of time what ratio of total available disk space gets assigned to each instance is not obvious as yet. The point about investigating multiple embedded RocksDB databases was to this end, i.e. if somehow both instances can be configured to compete for the same total space as before.
</blockquote></details>

The `maxSize` setting is only used for in-memory databases which are used for testing. Well, it looks like you might be able to set it for on-disk databases, but I'm not sure what we even do with the parameter in that instance. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KlyeVG2P82S8WB3WuSY:-KmDjlzbTEmE02_U3tAy:b-rb0yqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/11fa3e7d1a6790c9bfe8556db6bc0fd6f36af651/docs/RFCS/dedicated_raft_storage.md#L114)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Given my recent work on https://github.com/cockroachdb/cockroach/issues/14108, we'll definitely have to pay attention to how the separate RocksDB instances interact. For example, I seem to recall that we can configure them to use a shared background thread pool. Would that be better, or should they have separate thread pools for background compactions? 
</blockquote></details>

they can be configured to use a shared bg compaction thread pool yes, as for whether or not they should be I think it'll have to be determined experimentally. AFAICT the usage patterns here let us get away with very few compaction threads (we only need to compact what we truncate away, compactions need not happen frequently).

---

*[docs/RFCS/dedicated_raft_storage.md, line 284 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-Km-XOVoei9jnvpnWpgE:-KmDiUxl0WeEXHYzkF72:blj5aip) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db903f79474db493534f4fb66dfcabe3208b449/docs/RFCS/dedicated_raft_storage.md#L284)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `maxSize` setting is only used for in-memory databases which are used for testing. Well, it looks like you might be able to set it for on-disk databases, but I'm not sure what we even do with the parameter in that instance. 
</blockquote></details>

welp, you're right - apparently we do nothing in that case. `pkg/base/store_spec.go: StoreSpec.SizeInBytes` threw me off, gets passed down along from the cli (we can do `--store-path=/mnt/ssd01,size=0.02TiB`) but then isn't used to _size_ on-disk instances (used for when calculating 'free space' for re-balancing purposes).

Removed this whole section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 13 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRnIb-_P-IWaqj_OfR:-KmRnIb-_P-IWaqj_OfS:b8taniy) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L247)):*
> ```Markdown
> `raft` under our existing RocksDB storage directory.
> At the time of writing the keys that would need to be written to the new engine
> are the log keys and `HardState`.
> ```

We may want to consider using two column families for this, since the log keys are write-once and short-lived, while the hard state is overwritten frequently but never goes away completely. 

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmRmfnpJwd95tQ2sl2o:b-qflc66) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
> ```Markdown
> disk. If the node crashes at this point it will fail to load the
> `TruncatedState` and has no way to bridge the gap between the last persisted
> `ReplicaState` and the oldest entry in the truncated Raft log.
> ```

This seems bad. Do you have an idea for how to handle this? Maybe the truncation state should be stored in the log rocksdb instead of the regular one. 

---

*[docs/RFCS/dedicated_raft_storage.md, line 313 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRniN7fCGuewU3XK_S:-KmRniN7fCGuewU3XK_T:bsfysqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L313)):*
> ```Markdown
> migration at this time.
> 
> An online approach that could enable live migrations moving consensus state
> ```

We could also do a partially-online migration by using the new format for all new Replicas, so that as splits and rebalances occur we'd gradually move to the new format. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRnIb-_P-IWaqj_OfR:-KmRsGk8LrsbUK_QfADV:b-2l2okg) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L247)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We may want to consider using two column families for this, since the log keys are write-once and short-lived, while the hard state is overwritten frequently but never goes away completely. 
</blockquote></details>

No doubt you're aware, just wanted to point out explicitly that log keys are only *usually* write-once (log tail can be replaced after leadership change). If they were truly write once, we could use the RocksDB SingleDelete optimization, however useful that one may be.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmRqg3my51kMKv_sY48:b-gwdbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems bad. Do you have an idea for how to handle this? Maybe the truncation state should be stored in the log rocksdb instead of the regular one. 
</blockquote></details>

I think you just have to sync when you write the `TruncatedState`. Is that a big downside?

You can do better: you can only truncate the log once `TruncatedState` has been synced to disk. If we had a mechanism for finding out when that has happened, we could just remove the log entries then.

I don't know if it's relevant here, but could be related: we should refactor the way `evalTruncateLog` works. It currently takes writes all the way through the proposer-evaluated KV machinery, and at least from the graphs it looks that that's enough traffic to impair Raft throughput alone. We could lower the actual ranged clear below Raft (after all, no migration concerns there). We would be relaxing, somewhat, the stats which are now authoritative and would then only become "real" once the Raft log had actually been purged all the way up to the TruncatedState. I think there's no problem with that.

I haven't thought much about moving `TruncatedState` to the other engine. Seems possible too.

---

*[docs/RFCS/dedicated_raft_storage.md, line 313 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRniN7fCGuewU3XK_S:-KmRqBilQrsHpLb4rT6d:b-7jvymy) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L313)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We could also do a partially-online migration by using the new format for all new Replicas, so that as splits and rebalances occur we'd gradually move to the new format. 
</blockquote></details>

That's possible, but is it really a useful approach? We try to minimize data movement, and stable deployments could take forever to actually upgrade.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
thank you for the detailed reviews everyone! I don't see any outstanding comments so will be moving it to final comment period now.

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRnIb-_P-IWaqj_OfR:-KmRyZq4HMO2wmGqT3C2:bwygxsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L247)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No doubt you're aware, just wanted to point out explicitly that log keys are only *usually* write-once (log tail can be replaced after leadership change). If they were truly write once, we could use the RocksDB SingleDelete optimization, however useful that one may be.
</blockquote></details>

Noted.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmRorT3L0vq_z-Y9nLv:b6k0sgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
> Do you have an idea for how to handle this?

not yet, didn't get far enough into my experimentation branch to run into this. I have little context into the actual code implementation here but will mention alternatives considered when implementing (truncated state stored in the log RocksDB was one but I don't know how much internal coupling there is as yet).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-:-KmS-JLKKg1eIoG4sfzC:bh5ihgc)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmSPyDhwvF3F3cgI126:b-grw7x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

> Do you have an idea for how to handle this?

not yet, didn't get far enough into my experimentation branch to run into this. I have little context into the actual code implementation here but will mention alternatives considered when implementing (truncated state stored in the log RocksDB was one but I don't know how much internal coupling there is as yet).
</blockquote></details>

This section doesn't reflect the discussion here yet. I think you can just list these alternatives here and then decide on one when you're implementing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 1 files at r5.
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 247 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRnIb-_P-IWaqj_OfR:-KmSwFb9dOzth506LBXr:buqlgmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L247)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Noted.
</blockquote></details>

Ah, right. I was thinking about the naming scheme used for sideloaded sstables, which include the term and are really write-once. 

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmSwU-SJSBUG4ICOfbA:b-arv8lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
> I think you just have to sync when you write the TruncatedState. Is that a big downside?

This will be the only time we explicitly sync the KV rocksdb. It'll be an expensive sync and might cause performance problems. I think using the other engine will be better if it works. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmSxrm0jHL2CinV8tw2:b-txab7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> I think you just have to sync when you write the TruncatedState. Is that a big downside?

This will be the only time we explicitly sync the KV rocksdb. It'll be an expensive sync and might cause performance problems. I think using the other engine will be better if it works. 
</blockquote></details>

It's not just the `TruncatedState` that needs to be synced, but the associated effects of those entries, right? For example, if I `Put(a)` the primary RocksDB engine must have synced `a` before I can truncate that `Put` operation from the Raft log. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmSz6l_kVPqVmuEuFDP:b-l69xkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It's not just the `TruncatedState` that needs to be synced, but the associated effects of those entries, right? For example, if I `Put(a)` the primary RocksDB engine must have synced `a` before I can truncate that `Put` operation from the Raft log. 
</blockquote></details>

You're right, we need to explicitly sync whenever the synced AppliedIndex < new first log index.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmSyfANO3IGzmFQ4nwA:bek3aep) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You're right, we need to explicitly sync whenever the synced AppliedIndex < new first log index.
</blockquote></details>

Oops, I should have read the whole paragraph more closely. It's not just TruncatedState so moving that to the other engine probably won't help. In that case I think syncing the KV engine when writing TruncatedState is probably the best we can do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/dedicated_raft_storage.md, line 280 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361#-KmRmfnpJwd95tQ2sl2n:-KmT64aabY6ZJlKgB1xT:b-2qmqoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f7517a00c6ff04bea00c799eeb7cb6e9b4be94/docs/RFCS/dedicated_raft_storage.md#L280)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Oops, I should have read the whole paragraph more closely. It's not just TruncatedState so moving that to the other engine probably won't help. In that case I think syncing the KV engine when writing TruncatedState is probably the best we can do.
</blockquote></details>

Expanded the paragraph further for clarity and added takeaways from the discussion here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16361)*
<!-- Sent from Reviewable.io -->
